constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+
+Example 2
+
+Say that a user wants to construct a quadrilateral with ordered points A, B, C, and D such that ||CD|| is derived from ||AB|| and ||BC|| is derived from ||DA||. There are dependencies in both directions. Workarounds (1) and (2) above do not work since no single topological sort exists. The user _must_ factor out common expressions.
+
+Factoring out code seems like a workaround for a fundamental limitation of KCL, relative to existing CAD software. Even when constraints are possible in KCL today, it requires that users write them down in a very specific way that isn't necessarily intuitive, even using non-visual linguistic mathematical reasoning.
+# Proposal: A Constraint Language
+
+A constraint in the mechanical engineering world is a mathematical relationship between two expressions. The user should be able to write these relationships in any order. The language implementation should treat these relationships as a system of constraints and solve for whatever derived values that are required to construct geometry. Derived values should not need to be expressed as the outputs of functions with known inputs if their relationship is determined by an equation that can be symbolically rearranged to the same thing.
+
+As an example, we'll try to construct a sketch of a polygon where dependencies flow in both directions.
+
+Note that the code is generated by the user clicking in the UI. The resulting code is in the order that it was created, meaning it's a history of actions that the user performed in the UI.
+
+```
+p1 = screen_point(x: 10, y: 12)
+p2 = screen_point(x: 20, y: 13)
+l1 = line(p1, p2)
+p3 = screen_point(x: 21, y: 23)
+l2 = line(p2, p3)
+p4 = screen_point(x: 21, y: 23)
+l3 = line(p3, p4)
+l4 = line(p4, p1)
+||l3|| = 0.5 * ||l1||^2
+||l2|| = 1.2 * ||l4||
+```
+
+The above code can be read as follows:
+
+1. User enters sketch mode, which defaults to line drawing.
+2. User clicks a point. A variable `p1` is defined with the screen coordinates.
+3. User clicks another point. A variable `p2` is defined at that point. Since they're in line-sketch mode, `l1 = line(p1, p2)` is also generated. The app knows to name it with the prefix `l` because it is generating the line, so it knows it's a line. Similarly with points.
```

Comment:
How is construction geometry represented?

Reply:
If I understand correctly that construction geometry is geometry that is used only for building the model, but not intended to be exported or manufactured as part of the model, this language design currently doesn't treat it differently from other geometry. All geometry that has known coordinates will be displayed.

Perhaps we could add something similar to `hide`, but instead of not displaying, it omits the geometry from export.

Reply:
It's typically a UI toggle yeah, going from solid lines to dashed lines

Reply:
Yeah in Onshape's FeatureScript it is a function argument for geometry functions. For example, here is a construction point at the midpoint of a line segment:
```FeatureScript
skPoint(sketch, "BRe0F8pfMCRv.middle", { "construction" : true, "index" : "0" });

```